### PR TITLE
Pipeline-parallel support for Knowledge Distillation (NeMo 2)

### PR DESCRIPTION
**Pull request by AAnoosheh**
# What does this PR do ?

Enable Pipeline-Parallelism in conjunction with student-teacher distillation in NeMo 2

**Collection**: [LLM]

# Changelog 
- Create new script to enable KD in NeMo 2
- Add folder `nemo/collections/llm/distillation`
- Modify the `MegatronParallel` forward pass to run teacher in addition to student

# Usage

```bash
        python scripts/llm/gpt_distillation.py \
          --name experiment_name \
          --teacher_path /path/to/nemo2/teacher \
          --student_path /path/to/nemo2/student \
          --tp_size 4 \
          --cp_size 1 \
          --pp_size 2 \
          --devices 8 \
          --log_dir /tmp/nemo2_llama_distill \
          --max_steps 100 \
          --gbs 64 \
          --mbs 4 \
          --data_paths weight1 /path/to/data/1 weight2 /path/to/data/2 \
          --index_mapping_dir /path/to/data/cache/ \
          --seq_length 8192 \
          --warmup_steps 5 \
          --val_check_interval 50 \
          --log_interval 5
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

Megatron-LM vs NeMo 2.0 Llama3.1 8b->4b distillation comparison (Validation LM-Loss)
Screen Shot 2025-02-04 at 3 01 27 PM

### CODE REVIEW

File: nemo/collections/llm/distillation/model.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+    if isinstance(batch, tuple) and len(batch) == 3:
+        batch = batch[0]
+
+    required_device_keys = set()
+    required_host_keys = set()
+
+    if attn_mask_cpu:
+        # [ModelOpt]: We cache data for PP distillation, and save GPU mem by storing masks on CPU mem.
+        required_host_keys.add("attention_mask")
```
Comment (yashaswikarnati) [inline nemo/collections/llm/distillation/model.py:54]: I would double check this. I think we can pass this as None if its a regular causal mask and is defined in the spec.

File: scripts/llm/gpt_distillation.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        assert tokenizer is not None, "Please provide a model checkpoint with tokenizer included."
+
+    model = distill.DistillationGPTModel(
+        _student_model.config,
+        _teacher_model.config,
+        teacher_ckpt_path=args.teacher_path,
+        tokenizer=tokenizer,
+    )
+    # TODO(aanoosheh): Replace spec with modelopt one
+    model.__io__ = _student_model.__io__  # HACK: model saves and restores as original class
```
Comment (yashaswikarnati) [inline scripts/llm/gpt_distillation.py:108]: wonder why do we need this

File: scripts/llm/gpt_distillation.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        log_every_n_steps=args.log_interval,
+        val_check_interval=args.val_check_interval,
+        limit_val_batches=args.limit_val_batches,
+        strategy=strategy,
+        accelerator="gpu",
+        plugins=nl.MegatronMixedPrecision(precision=args.precision),
+    )
+
+    ## Load both models and combine into an aggregate module
+    if args.cicd_run:
```
Comment (yashaswikarnati) [inline scripts/llm/gpt_distillation.py:90]: For CICD, wonder if its better to separate out logic and have a separate script in tests folder.

File: scripts/llm/gpt_distillation.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    parser.add_argument("--warmup_steps", type=int, default=50, help="""Number of scheduler warmup steps""")
+    parser.add_argument("--val_check_interval", type=int, default=100, help="""Run validation every _ steps""")
+    parser.add_argument("--limit_val_batches", type=int, default=32, help="""Number of batches per validation stage""")
+    parser.add_argument("--log_interval", type=int, default=10, help="""Write to log every _ steps""")
+    parser.add_argument("--cicd_run", action="store_true", help="Used only when called by NeMo CI")
+
+    return parser.parse_args()
+
+
+########################################################
```
Comment (yashaswikarnati) [inline scripts/llm/gpt_distillation.py:64]: nit; remove

File: nemo/collections/llm/distillation/utils.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                if hasattr(model_wrapper, "_config"):
+                    model_wrapper.config = model_wrapper._config
+                    del model_wrapper._config
+
+        # HACK: Pipeline-parallel forward function relies on the config in the model to know what
+        # hidden size of tensor to communicate to next stage.
+        model.swap_teacher_config = MethodType(_swap_teacher_config, model)
+
+
+def load_cicd_models(student_path: str):
```
Comment (yashaswikarnati) [inline nemo/collections/llm/distillation/utils.py:157]: I would move cicd related stuff to separate script in tests folder

File: nemo/collections/llm/distillation/utils.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            self.it = iter(self.data)
+            return next(self.it)
+
+
+def adjust_distillation_model_for_mcore(
+    model: mtd.DistillationModel, model_cfg: "TransformerConfig", distill_cfg: Dict[str, Any]
+):
+    """Extra modifcations to ``mtd.DistillationModel`` requried for Megatron-Core."""
+
+    # HACK: Get rid of ModelOpt Distillation state
```
Comment (yashaswikarnati) [inline nemo/collections/llm/distillation/utils.py:101]: nit; please remove #HACK in below

File: nemo/collections/llm/distillation/model.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    def core_module(self):
+        # pylint: disable=C0116
+        return unwrap_model(self.module, (DDP, Float16Module, MCoreFloat16Module))
+
+    def train(self, mode: bool = True):
+        # pylint: disable=C0116
+        self._train_called = True
+        return super().train(mode)
+
+    def __setattr__(self, name, value):
```
Comment (yashaswikarnati) [inline nemo/collections/llm/distillation/model.py:230]: wonder why we need this here, not sure i understand

File: nemo/collections/llm/distillation/model.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+            torch.distributed.all_reduce(loss, group=parallel_state.get_tensor_model_parallel_group())
+
+        return loss
+
+
+class DistillationGPTModel(llm.GPTModel):
+    """Custom GPT subclass for distillation-related modifications."""
+
+    def __init__(
+        self,
```
Comment (yashaswikarnati) [inline nemo/collections/llm/distillation/model.py:136]: I would add doc strings for this class as this is what we expose to user

File: nemo/collections/llm/distillation/model.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        else:
+            batch_required_keys[key] = None
+
+    # slice batch along sequence dimension for context parallelism
+    output = get_batch_on_this_context_parallel_rank(batch_required_keys)
+
+    return output
+
+
+class _DistillationLossReduction(MaskedTokenLossReduction):
```
Comment (yashaswikarnati) [inline nemo/collections/llm/distillation/model.py:83]: nit; may be move to loss.py

File: nemo/collections/llm/distillation/model.py (hunk 1 above)
Comment (yashaswikarnati) [reply nemo/collections/llm/distillation/model.py:54]: https://github.com/NVIDIA/NeMo/blob/b9bdd451afa944eda50aed8922414bc133a8e6d3/nemo/collections/llm/gpt/data/pre_training.py#L153

attention mask creation defaults to False. If not using TE this is needed though

https://github.com/NVIDIA/NeMo/blob/b9bdd451afa944eda50aed8922414bc133a8e6d3/nemo/collections/llm/gpt/data/pre_training.py#L199

File: nemo/collections/llm/distillation/model.py (hunk 7 above)
Comment (AAnoosheh) [reply nemo/collections/llm/distillation/model.py:230]: It's a bit crazy, but after sanity checking phase is done, Pytorch Lightning will directly set `module.training = True` instead of calling `module.train()` so we need to eventually get them to change it on their end.

File: nemo/collections/llm/distillation/model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AAnoosheh) [reply nemo/collections/llm/distillation/model.py:54]: But before it would have gone to the `val.cuda()` path. Should we fix both?

[the remark this answers is not in the document]
Comment (AAnoosheh) [reply nemo/collections/llm/distillation/model.py:54]: (This is copied and modified directly from `nemo/collections/llm/gpt/model/base.py`)

File: nemo/collections/llm/distillation/model.py (hunk 9 above)
Comment (AAnoosheh) [reply nemo/collections/llm/distillation/model.py:83]: I think it belongs better here, since loss.py is currently model-agnostic

File: nemo/collections/llm/distillation/model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AAnoosheh) [reply nemo/collections/llm/distillation/model.py:54]: I think actually it's not None, but just won't exist in the dict at all.  In Megatron-LM it can be None though.

File: .github/workflows/cicd-main.yml [hunk 10]
```diff
@@ -4949,6 +4949,36 @@ jobs:
         rm -rf /tmp/nemo2_ckpt
         rm -rf /tmp/nemo2_ptq_engine
 
+  L2_NeMo_2_Distill_Llama3_TP1PP2:
+    needs: [pre-flight, cicd-test-container-build]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.pre-flight.outputs.test_to_run), 'L2_NeMo_2_Distill_Llama3_TP1PP2') || needs.pre-flight.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:4957]: Out of curiosity: Will this test require 1 or two gpus? For a single, please use `self-hosted-azure-gpus-1` and for dual the current tag is good

Comment (yashaswikarnati) [reply .github/workflows/cicd-main.yml:4957]: we would need dual to test PP>1

**Review by ko3n1g — DISMISSED** [review]
Makes sense, thanks!

**Review by yashaswikarnati — DISMISSED** [review]

**Review by yashaswikarnati — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]
